scripts/run_all.py

```diff
@@ -4,86 +4,239 @@
 
 import argparse
 import os
+import socket
 import subprocess
 import time
 from glob import glob
-from typing import Any, List
+from types import TracebackType
+from typing import Any, List, Optional, Tuple, Type, cast
 
 
-def run_py_example(path: str, args: List[str] = []) -> None:
+def run_py_example(path: str, viewer_port: int, wait: bool = True, save: Optional[str] = None) -> Any:
+    args = ["--connect", "--addr", f"127.0.0.1:{viewer_port}"]
+    if save is not None:
+        args += ["--save", str(save)]
+
     process = subprocess.Popen(
         ["python3", "main.py", "--num-frames=30", "--steps=200"] + args,
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
+    return process
 
 
-def run_saved_example(path: str, args: List[str] = []) -> None:
+def run_saved_example(path: str, wait: bool = True) -> Any:
     process = subprocess.Popen(
-        ["cargo", "run", "-p", "rerun", "--all-features", "--", "out.rrd"] + args,
+        ["cargo", "run", "-p", "rerun", "--all-features", "--", "out.rrd"],
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
+    return process
 
 
-def collect_examples() -> List[str]:
-    return [os.path.dirname(entry) for entry in glob("examples/python/**/main.py")]
+def get_free_port() -> int:
+    with socket.socket() as s:
+        s.bind(("", 0))
+        return cast(int, s.getsockname()[1])
```

Comment:
```suggestion
        return int(getsockname()[1])
```

…I think it is the same thing and more pythonic

Reply:
Yes, I wasn't sure what the best way was. I figured since it's already an int, I just have to coerce the type checker to believe that. But an explicit cast to int is much better. Updated.